File: client/path_cache.go
package client

import (
	"github.com/hashicorp/golang-lru/v2"
)

// PathCacheEvictHandler is called when a cursor is evicted from the cache.
type PathCacheEvictHandler func(*PathCursor)

// PathCache implements a mutual path aliasing scheme with a server.
type PathCache struct {
	cache   *lru.Cache[uint32, *PathCursor]
	evictCb PathCacheEvictHandler
}

// NewPathCache builds a path cache.
func NewPathCache(size int, evictCb PathCacheEvictHandler) *PathCache {
	c := &PathCache{evictCb: evictCb}
	cache, err := lru.NewWithEvict(size, c.onEvicted)
	if err != nil {
		panic(err)
	}
	c.cache = cache
	return c
}

// Get returns the value with the key.
func (p *PathCache) Get(key uint32) *PathCursor {
	val, ok := p.cache.Get(key)
	if !ok {
		return nil
	}
	return val
}

// Set sets a value in the cache.
func (p *PathCache) Set(key uint32, val *PathCursor) {
	p.cache.Add(key, val)
}

// onEvicted handles when a key is evicted from the cache.
func (p *PathCache) onEvicted(key uint32, value *PathCursor) {
	if p.evictCb == nil {
		return
	}
	p.evictCb(value)
}

File: src/result-tree/path-cache.ts
import { PathCursor } from './path-cursor'
import { LRUMap } from 'lru_map'

// PathCacheEvictHandler is called when a cursor is evicted from the cache.
export type PathCacheEvictHandler = (cursor: PathCursor) => void

// PathCache implements a mutual path aliasing scheme with a server.
export class PathCache {
  // cache contains cached path cursors
  private cache: LRUMap<number, PathCursor>
  // evictCb is the eviction callback
  private evictCb: PathCacheEvictHandler | null

  constructor(size: number, evictCb: PathCacheEvictHandler | null) {
    this.evictCb = evictCb
    this.cache = new LRUMap<number, PathCursor>(size)
  }

  // get retrieves a cursor by key, registering use
  public get(key: number): PathCursor | null {
    return this.cache.get(key) || null
  }

  // reset resets the cache
  public reset() {
    this.cache.clear()
  }

  // set sets the cursor by key and value
  public set(key: number, val: PathCursor) {
    this.setWithEvict(key, val)
  }

  // setWithEvict sets the key to the value.
  // the map is checked to see if any values will be evicted.
  // if so, the evicted value is shifted first and emitted to the callback
  private setWithEvict(key: number, value: PathCursor) {
    let cacheSize = this.cache.size
    let cacheLimit = this.cache.limit
    let exists = this.cache.has(key)
    if (cacheSize === cacheLimit && !exists) {
      let removedTuple = this.cache.shift()
      if (this.evictCb && removedTuple) {
        this.evictCb(removedTuple[1])
      }
    }
    this.cache.set(key, value)
  }
}
